File: Design/1396. Design Underground System/main.go
package main

import "fmt"

// UndergroundSystem ...
type UndergroundSystem struct {
	station map[string]map[int]int
}

// Constructor ...
func Constructor() UndergroundSystem {
	obj := make(map[string]map[int]int)
	return UndergroundSystem{station: obj}
}

// CheckIn A customer with id card equal to id, gets in the station stationName at time t.
func (us *UndergroundSystem) CheckIn(id int, stationName string, t int) {
	if us.station[stationName] == nil {
		us.station[stationName] = make(map[int]int)
	}
	us.station[stationName][id] = t
}

// CheckOut A customer with id card equal to id, gets out from the station stationName at time t.
func (us *UndergroundSystem) CheckOut(id int, stationName string, t int) {
	if us.station[stationName] == nil {
		us.station[stationName] = make(map[int]int)
	}
	us.station[stationName][id] = t
}

// GetAverageTime Returns the average time to travel between the startStation and the endStation.
func (us *UndergroundSystem) GetAverageTime(startStation string, endStation string) float64 {
	count, sum := 0, 0
	start := us.station[startStation]
	end := us.station[endStation]
	for id, t := range start {
		if end[id] != 0 {
			sum = sum - t + end[id]
			count++
		}
	}
	return float64(sum) / float64(count)
}

func main() {
	undergroundSystem := Constructor1()
	undergroundSystem.CheckIn1(10, "Leyton", 3)
	undergroundSystem.CheckOut1(10, "Paradise", 8)
	fmt.Println("Output: ", undergroundSystem.GetAverageTime1("Leyton", "Paradise")) // return 5.00000
	undergroundSystem.CheckIn1(5, "Leyton", 10)
	undergroundSystem.CheckOut1(5, "Paradise", 16)
	fmt.Println("Output: ", undergroundSystem.GetAverageTime1("Leyton", "Paradise")) // return 5.50000
	undergroundSystem.CheckIn1(2, "Leyton", 21)
	undergroundSystem.CheckOut1(2, "Paradise", 30)
	fmt.Println("Output: ", undergroundSystem.GetAverageTime1("Leyton", "Paradise")) // return 6.66667
}

// Station ...
type Station struct {
	stationName string
	time        int
}

// UndergroundSystem1 ...
type UndergroundSystem1 struct {
	in  map[int]*Station
	out map[string][]int
}

// Constructor1 ...
func Constructor1() UndergroundSystem1 {
	us := new(UndergroundSystem1)
	us.in = map[int]*Station{}
	us.out = map[string][]int{}
	return *us
}

// CheckIn1 A customer with id card equal to id, gets in the station stationName at time t.
func (us *UndergroundSystem1) CheckIn1(id int, stationName string, t int) {
	s := &Station{stationName, t}
	us.in[id] = s
}

// CheckOut1 A customer with id card equal to id, gets out from the station stationName at time t.
func (us *UndergroundSystem1) CheckOut1(id int, stationName string, t int) {
	s := us.in[id]
	name := s.stationName + stationName
	time := t - s.time
	out := us.out[name]
	out = append(out, time)
	us.out[name] = out
}

// GetAverageTime1 Returns the average time to travel between the startStation and the endStation.
func (us *UndergroundSystem1) GetAverageTime1(startStation string, endStation string) float64 {
	name := startStation + endStation
	out := us.out[name]
	sum := 0
	for _, t := range out {
		sum = sum + t
	}
	return float64(sum) / float64(len(out))
}

File: Design/1396. Design Underground System/main.c
#include <stdio.h>
#include <stdlib.h>
#include "uthash.h"

typedef struct {
    int key;
    char in[20];
    int time;
    UT_hash_handle hh;
} Passenger;

typedef struct {
    char in[20];
    char out[20];
} StartEndHash;

typedef struct {
    StartEndHash key;
    int time[10000];
    int count;
    UT_hash_handle hh;
} Stations;

typedef struct {
    Passenger *pPassenger;
    Stations *pStations;
} UndergroundSystem;


UndergroundSystem* undergroundSystemCreate() {
    UndergroundSystem *pUS = malloc(sizeof(UndergroundSystem));
    pUS->pPassenger = NULL;
    pUS->pStations = NULL;
    return pUS;
}

void undergroundSystemCheckIn(UndergroundSystem* obj, int id, char * stationName, int t) {
    Passenger *p = NULL;
    HASH_FIND_INT(obj->pPassenger, &id, p);
    if (p == NULL) {
        p = malloc(sizeof(Passenger));
        p->key = id;
        HASH_ADD_INT(obj->pPassenger, key, p);
    }
    strcpy(p->in, stationName);
    p->time = t;
}

void undergroundSystemCheckOut(UndergroundSystem* obj, int id, char * stationName, int t) {
    Passenger *p = NULL;
    Stations *s = NULL;
    StartEndHash key;

    HASH_FIND_INT(obj->pPassenger, &id, p);
    if (p == NULL) {
        return;
    }
    
    memset(&key, 0, sizeof(key));
    strcpy(key.in, p->in);
    strcpy(key.out, stationName);

    HASH_FIND(hh, obj->pStations, &key, sizeof(key), s);
    if (s == NULL) {
        s = malloc(sizeof(Stations));
        memcpy(&s->key, &key, sizeof(key));
        s->count = 0;
        HASH_ADD(hh, obj->pStations, key, sizeof(key), s);
    }

    s->time[s->count++] = t - p->time;
}

double undergroundSystemGetAverageTime(UndergroundSystem* obj, char * startStation, char * endStation) {
    Stations *s = NULL;
    double sum = 0;
    StartEndHash key;

    memset(&key, 0, sizeof(key));
    strcpy(key.in, startStation);
    strcpy(key.out, endStation);
    HASH_FIND(hh, obj->pStations, &key, sizeof(key), s);
    if (s == NULL) {
        return 0.0;
    }
    for (int i = 0; i < s->count; i++) {
        sum += s->time[i];
    }
    return sum / s->count;
}

void undergroundSystemFree(UndergroundSystem* obj) {
    Stations *s = NULL, *ss;
    Passenger *p = NULL, *pp;
    HASH_ITER(hh, obj->pStations, s, ss) {
        HASH_DEL(obj->pStations, s);
        free(s);
    }
    HASH_ITER(hh, obj->pPassenger, p, pp) {
        HASH_DEL(obj->pPassenger, p);
        free(p);
    }
    free(obj);
}

/**
 * Your UndergroundSystem struct will be instantiated and called as such:
 * UndergroundSystem* obj = undergroundSystemCreate();
 * undergroundSystemCheckIn(obj, id, stationName, t);
 
 * undergroundSystemCheckOut(obj, id, stationName, t);
 
 * double param_3 = undergroundSystemGetAverageTime(obj, startStation, endStation);
 
 * undergroundSystemFree(obj);
*/
int main() {
    UndergroundSystem* pUS =  undergroundSystemCreate();
    undergroundSystemCheckIn(pUS, 45, "Leyton", 3);
    undergroundSystemCheckIn(pUS, 32, "Paradise", 8);
    undergroundSystemCheckIn(pUS, 27, "Leyton", 10);
    undergroundSystemCheckOut(pUS, 45, "Waterloo", 15);
    undergroundSystemCheckOut(pUS, 27, "Waterloo", 20);
    undergroundSystemCheckOut(pUS, 32, "Cambridge", 22);

    double param_3 = undergroundSystemGetAverageTime(pUS, "Leyton", "Waterloo");
    printf("Output: %f.", param_3);

    undergroundSystemFree(pUS);
    return 0;
}